File: ccdr/makeitso/iOS/MakeItSo/MakeItSo/Services/AITaskService.swift
import Foundation
import FirebaseAILogic

@MainActor
class AITaskService {
  private let model: GenerativeModel
  
  init() {
    // 1. Define schema for a JSON array of strings
    let responseSchema = Schema.array(
      items: .string(description: "A single actionable, independent, top-level task to be done")
    )
    
    let config = GenerationConfig(
      responseMIMEType: "application/json",
      responseSchema: responseSchema
    )
    
    // 2. Initialize Firebase AI Logic with the Google AI Developer API backend
    let ai = FirebaseAI.firebaseAI(backend: .googleAI())
    
    // 3. Instantiate with the latest modern model: gemini-2.5-flash
    self.model = ai.generativeModel(
      modelName: "gemini-2.5-flash",
      generationConfig: config
    )
  }
  
  func breakDownTask(title: String) async throws -> [String] {
    let prompt = """
    Break down the following complex task into exactly 4-6 simple, actionable, and concrete top-level tasks.
    
    Task: "\(title)"
    """
    
    let response = try await model.generateContent(prompt)
    guard let responseText = response.text else {
      throw NSError(domain: "AITaskService", code: 404, userInfo: [NSLocalizedDescriptionKey: "No response from AI"])
    }
    
    guard let data = responseText.data(using: .utf8) else { return [] }
    return try JSONDecoder().decode([String].self, from: data)
  }
}

File: ccdr/makeitso/Android/MakeItSo/app/src/main/java/com/google/firebase/example/makeitso/data/repository/AITaskService.kt
package com.google.firebase.example.makeitso.data.repository

import com.google.firebase.Firebase
import com.google.firebase.ai.ai
import com.google.firebase.ai.type.GenerativeBackend
import com.google.firebase.ai.type.generationConfig
import com.google.firebase.ai.type.Schema
import kotlinx.serialization.json.Json
import kotlinx.serialization.decodeFromString
import javax.inject.Inject

class AITaskService @Inject constructor() {
    private val responseSchema = Schema.array(
        items = Schema.string(description = "A single actionable, independent, top-level task to be done")
    )
    
    private val config = generationConfig {
        responseMimeType = "application/json"
        responseSchema = this@AITaskService.responseSchema
    }
    
    // Initialize Firebase AI Logic with the Google AI Developer API backend and gemini-2.5-flash
    private val model = Firebase.ai(backend = GenerativeBackend.googleAI()).generativeModel(
        modelName = "gemini-2.5-flash",
        generationConfig = config
    )
    
    suspend fun breakDownTask(title: String): List<String> {
        val prompt = """
            Break down the following complex task into exactly 4-6 simple, actionable, and concrete top-level tasks.
            
            Task: "$title"
        """.trimIndent()
        
        val response = model.generateContent(prompt)
        val responseText = response.text ?: return emptyList()
        
        return try {
            Json.decodeFromString<List<String>>(responseText)
        } catch (e: Exception) {
            emptyList()
        }
    }
}
